### PR TITLE
Added Memory Requests Is Not Defined query #491

### LINKED ISSUE #491: Add Memory Requests Is Not Defined query for Kubernetes
### Platform
Kubernetes

### Description
It's important to specify a memory request.

### PR DESCRIPTION
**Pull request by joelcarvalhocheckmarx**
It's important to specify a memory request.

Closes #491

### CODE REVIEW

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                "issueType":		   "MissingAttribute",
+                "keyExpectedValue": "spec.containers are defined",
+                "keyActualValue": 	"spec.containers are undefined"
+              }
+}
+
+CxPolicy [ result ] {
+   spec := input.document[i].spec
+   exists_containers := object.get(spec, "containers", "undefined") != "undefined"
+   exists_containers
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:36]: redundant can be removed

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                "issueType":		   "MissingAttribute",
+                "keyExpectedValue": sprintf("spec.containers[%d].resources are defined", [c]),
+                "keyActualValue": 	sprintf("spec.containers[%d].resources are undefined", [c])
+              }
+}
+
+CxPolicy [ result ] {
+   spec := input.document[i].spec
+   exists_containers := object.get(spec, "containers", "undefined") != "undefined"
+   exists_containers
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:54]: can be removed is redudant

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+   
+   containers := spec.containers
+   exists_resources := object.get(containers[c], "resources", "undefined") != "undefined"
+   exists_resources
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:58]: can be removed is redudant

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 4]
```diff
@@ -0,0 +1,71 @@
+package Cx
+
+CxPolicy [ result ] {
+   spec := input.document[i].spec
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:4]: metadata := document.metadata
spec := document.spec 
containers := spec.containers

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package Cx
+
+CxPolicy [ result ] {
+   spec := input.document[i].spec 
+   requests := spec.containers[c].resources.requests
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:5]: requests := containers[c].resources.requests

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+   exists_memory := object.get(requests, "memory", "undefined") != "undefined"
+   not exists_memory
+   
+   result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("spec.containers[%d].requests.memory", [c]),
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:11]: "searchKey": 	    sprintf("metadata.name=%s.spec.containers.name=%s.requests.memory", [metadata.name, containers[c].name]),

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+
+CxPolicy [ result ] {
+   spec := input.document[i].spec
+   exists_containers := object.get(spec, "containers", "undefined") != "undefined"
+   not exists_containers
+   
+   result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    "spec",
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:26]: "searchKey": sprintf("metadata.name=%s.spec", [metadata.name]),

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package Cx
+
+CxPolicy [ result ] {
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:3]: document := input.document[i]

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+   exists_containers := object.get(spec, "containers", "undefined") != "undefined"
+   exists_containers
+   
+   containers := spec.containers
+   exists_resources := object.get(containers[c], "resources", "undefined") != "undefined"
+   not exists_resources
+   
+   result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("spec.containers[%d]", [c]),
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:44]: "searchKey": sprintf("metadata.name=%s.spec.containers.name=%s", [metadata.name, containers[c].name]),

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+   exists_resources := object.get(containers[c], "resources", "undefined") != "undefined"
+   exists_resources
+   
+   resources := spec.containers[c].resources
+   exists_requests := object.get(resources, "requests", "undefined") != "undefined"
+   not exists_requests
+   
+   result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	    sprintf("spec.containers[%d]", [c]),
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:66]: "searchKey": sprintf("metadata.name=%s.spec.containers.name=%s", [metadata.name, containers[c].name]),

File: assets/queries/k8s/memory_requests_not_defined/test/positive_expected_result.json [hunk 11]
```diff
@@ -0,0 +1,7 @@
+[
+	{
+		"queryName": "Memory Requests Is Not Defined",
+		"severity": "MEDIUM",
+		"line": 6
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/test/positive_expected_result.json:5]: "line": 13

**Review by joelsou5a — CHANGES_REQUESTED** [review]

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                "searchKey": 	    sprintf("spec.containers[%d].requests.memory", [c]),
+                "issueType":		   "MissingAttribute",
+                "keyExpectedValue": sprintf("spec.containers[%d].resources.requests.memory is defined", [c]),
+                "keyActualValue": 	sprintf("spec.containers[%d].resources.requests.memory is undefined", [c])
+              }
+}
+
+
+CxPolicy [ result ] {
+   spec := input.document[i].spec
```
Comment (joelsou5a) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:20]: document := input.document[i]
 spec := document.spec
metadata := document.metadata

**Review by joelsou5a — CHANGES_REQUESTED** [review]

File: assets/queries/k8s/memory_requests_not_defined/query.rego [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package Cx
+
+CxPolicy [ result ] {
+   metadata := input.document[i].metadata
+   spec := input.document[i].spec 
+   containers := spec.containers
+   requests := containers[c].resources.requests
```
Comment (joaomartinscx) [inline assets/queries/k8s/memory_requests_not_defined/query.rego:7]: The query won't detect that these attributes, *resources* and *requests*, are not present. If you wish to enforce an attribute to existing at a lower level, you should guarantee the parent exists. I suggest to test these different edge cases and present different results for each.

**Review by joaomartinscx — DISMISSED** [review]

**Review by joelsou5a — APPROVED** [review]
